File: solune/frontend/src/types/workflow.ts
/**
 * Workflow domain types — agent assignment, workflow configuration, notifications.
 */

import type { ResolvedModelInfo } from './common';

// ============ Agent Assignment (004-agent-workflow-config-ui) ============

export type AgentSource = 'builtin' | 'repository';

export interface AgentAssignment {
  id: string; // UUID string
  slug: string; // Agent identifier
  display_name?: string | null;
  config?: Record<string, unknown> | null;
}

export interface AvailableAgent {
  slug: string;
  display_name: string;
  description?: string | null;
  avatar_url?: string | null;
  icon_name?: string | null;
  default_model_id?: string;
  default_model_name?: string;
  tools_count?: number | null;
  source: AgentSource;
}

export interface AgentPreset {
  id: string;
  label: string;
  description: string;
  mappings: Record<string, AgentAssignment[]>;
}

// ============ Workflow Result (T052) ============

export interface WorkflowResult {
  success: boolean;
  issue_id?: string;
  issue_number?: number;
  issue_url?: string;
  project_item_id?: string;
  current_status?: string;
  message: string;
  resolved_model?: ResolvedModelInfo | null;
}

export interface PipelineIssueLaunchRequest {
  issue_description: string;
  pipeline_id: string;
}

export interface WorkflowConfiguration {
  project_id: string;
  repository_owner: string;
  repository_name: string;
  copilot_assignee: string;
  review_assignee?: string;
  agent_mappings: Record<string, AgentAssignment[]>;
  status_backlog: string;
  status_ready: string;
  status_in_progress: string;
  status_in_review: string;
  enabled: boolean;
}

export interface AgentNotification {
  type: 'agent_assigned' | 'agent_completed';
  issue_number: number;
  agent_name: string;
  status: string;
  next_agent: string | null;
  timestamp: string;
}

export interface PipelineStateInfo {
  issue_number: number;
  project_id: string;
  status: string;
  agents: string[];
  current_agent_index: number;
  current_agent: string | null;
  completed_agents: string[];
  is_complete: boolean;
  started_at: string | null;
  error: string | null;
  queued: boolean;
}

File: solune/backend/src/models/workflow.py
"""Workflow configuration, transition audit, and result models."""

from datetime import datetime
from enum import StrEnum
from uuid import UUID, uuid4

from pydantic import BaseModel, Field, field_validator

from src.constants import AGENT_DISPLAY_NAMES, DEFAULT_AGENT_MAPPINGS
from src.models.agent import AgentAssignment, AgentAssignmentInput
from src.utils import utcnow


class ExecutionGroupMapping(BaseModel):
    """A group of agents within a workflow status sharing an execution mode.

    Created during the conversion from ``PipelineConfig`` →
    ``WorkflowConfiguration`` to preserve execution-group structure
    through the orchestration layer.
    """

    group_id: str = Field(..., description="Unique group identifier")
    order: int = Field(default=0, description="Execution order within the stage")
    execution_mode: str = Field(default="sequential", description="'sequential' or 'parallel'")
    agents: list[AgentAssignment] = Field(
        default_factory=list[AgentAssignment], description="Ordered agent assignments in this group"
    )

    @field_validator("execution_mode")
    @classmethod
    def validate_execution_mode(cls, v: str) -> str:
        if v not in ("sequential", "parallel"):
            raise ValueError("execution_mode must be 'sequential' or 'parallel'")  # noqa: TRY003 — reason: domain exception with descriptive message
        return v


class TriggeredBy(StrEnum):
    """Source that triggered a workflow transition."""

    AUTOMATIC = "automatic"
    MANUAL = "manual"
    DETECTION = "detection"


class WorkflowConfiguration(BaseModel):
    """Configuration for the workflow orchestrator."""

    project_id: str = Field(..., description="GitHub Project node ID")
    repository_owner: str = Field(..., description="Target repository owner")
    repository_name: str = Field(..., description="Target repository name")
    copilot_assignee: str = Field(
        default="", description="Username for implementation (empty to skip assignment)"
    )
    review_assignee: str | None = Field(
        default=None, description="Username for review (default: repo owner)"
    )
    agent_mappings: dict[str, list[AgentAssignmentInput]] = Field(
        default_factory=lambda: {
            k: [
                AgentAssignment(
                    slug=s,
                    display_name=AGENT_DISPLAY_NAMES.get(s),
                )
                for s in v
            ]
            for k, v in DEFAULT_AGENT_MAPPINGS.items()
        },
        description="Status name → ordered list of agent assignments",
    )
    status_backlog: str = Field(default="Backlog", description="Backlog status column name")
    status_ready: str = Field(default="Ready", description="Ready status column name")
    status_in_progress: str = Field(default="In Progress", description="In Progress column name")
    status_in_review: str = Field(default="In Review", description="In Review column name")
    enabled: bool = Field(default=True, description="Whether workflow automation is active")
    stage_execution_modes: dict[str, str] = Field(
        default_factory=dict,
        description="Status name → execution mode ('sequential' | 'parallel')",
    )
    group_mappings: dict[str, list[ExecutionGroupMapping]] = Field(
        default_factory=dict,
        description="Status name → ordered list of execution groups. Empty for legacy pipelines.",
    )


class WorkflowTransition(BaseModel):
    """Audit log for workflow status transitions."""

    transition_id: UUID = Field(default_factory=uuid4, description="Unique transition ID")
    issue_id: str = Field(..., description="GitHub Issue node ID")
    project_id: str = Field(..., description="GitHub Project node ID")
    from_status: str | None = Field(default=None, description="Previous status (null for initial)")
    to_status: str = Field(..., description="New status")
    assigned_user: str | None = Field(default=None, description="User assigned (if applicable)")
    triggered_by: TriggeredBy = Field(..., description="Transition trigger source")
    success: bool = Field(..., description="Whether transition succeeded")
    error_message: str | None = Field(default=None, description="Error details if failed")
    timestamp: datetime = Field(default_factory=utcnow, description="Transition timestamp")


class WorkflowResult(BaseModel):
    """Result of a workflow operation (confirm/reject)."""

    success: bool = Field(..., description="Whether operation succeeded")
    issue_id: str | None = Field(default=None, description="GitHub Issue node ID")
    issue_number: int | None = Field(default=None, description="Human-readable issue number")
    issue_url: str | None = Field(default=None, description="URL to issue on GitHub")
    project_item_id: str | None = Field(default=None, description="GitHub Project item ID")
    current_status: str | None = Field(default=None, description="Current workflow status")
    message: str = Field(..., description="Human-readable result message")


__all__ = [
    "ExecutionGroupMapping",
    "TriggeredBy",
    "WorkflowConfiguration",
    "WorkflowResult",
    "WorkflowTransition",
]
